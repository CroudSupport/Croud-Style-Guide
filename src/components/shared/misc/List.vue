<template>
    <div class="switcher">
        <div v-if="showList" ref="universalList" :class="dropdownClasses">
            <span class="text">{{ buttonText }}</span>
            <i v-if="showIcon" class="icon" :class="buttonIcon"></i>

            <div class="menu">
                <slot name="custom-header">
                    <div v-if="showHeader && showHeaderAction" class="header-action">
                        <slot name="header-action"></slot>
                    </div>

                    <div v-if="showHeader" class="header">
                        
                        <slot name="header-item">
                            <div class="item">
                                {{ getHeaderItem.name }}
                            </div>
                        </slot>

                        <div v-if="!getHeaderItem.name" class="item">
                            <slot name="default-header-item">
                                Select an item
                            </slot>
                        </div>
                    </div>
                </slot>
                
                <div v-if="showHeader" class="divider"></div>
            
                <slot name="extra-items">
                </slot>

                <div class="scrolling menu">
                    <div v-for="item in list" :key="item.id" class="item" @click="$emit('item-selected', item)">
                        <slot name="items" :item="item">
                            {{ item.name }}
                        </slot>
                    </div>
                     
                    <div v-if="!list.length" class="item">
                        <slot name="list-empty">
                            No items to display
                        </slot>
                    </div>
                </div>

                <div class="ui inverted dimmer" :class="{active: loading}">
                    <div class="ui text large loader">{{ loadingText }}</div>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
    /**
    * An universal list dropdown component
    *
    * @example ./croud-list-dropdown.md
    */

    export default {
        name: 'croud-list-dropdown',

        props: {
            /**
            * Show universalListDropdown
            */
            showList: {
                type: Boolean,
                default: true,
            },

            /**
            * Show loader in menu
            */
            loading: {
                type: Boolean,
                default: false,
            },

            /**
            * Customize the loading text
            */
            loadingText: {
                type: String,
                default: 'Loading',
            },

            /**
            * Show header
            */
            showHeader: {
                type: Boolean,
                default: true,
            },

            /**
            * Show header action
            */
            showHeaderAction: {
                type: Boolean,
                default: false,
            },

            /**
            * Header item can show the currently selected item, just pass it to this prop
            */
            headerItem: {
                default() {
                    return {}
                },
            },

            /**
            * Pass the getter for your list in this prop
            */
            listGetter: {
                default() {
                    return []
                },
            },

            /**
            * Semantic-ui dropdown overides
            */
            dropdownSettings: {
                default() {
                    return {
                        forceSelection: false,
                        action: 'select',
                    }
                },
            },

            /**
            * Dropdown classes
            */
            dropdownClasses: {
                default: 'ui floating top right pointing dropdown tiny positive button right floated',
            },

            /**
            * Show icon in dropdown button
            */
            showIcon: {
                default() {
                    return false
                },
            },

            /**
            * Text shown in the dropdown button
            */
            buttonText: {
                default() {
                    return 'Switch'
                },
            },

            /**
            * Icon displayed in the dropdown button (requires 'icon' adding to dropdown classes)
            */
            buttonIcon: {
                default() {
                    return ''
                },
            },
        },

        methods: {
            buildDropdown() {
                $(this.$refs.universalList).dropdown(this.dropdownSettings)
            },
        },

        computed: {
            list() {
                return this.listGetter || []
            },

            getHeaderItem() {
                return this.headerItem || {}
            },
        },

        mounted() {
            this.buildDropdown()
        },

        watch: {
            showSwitcher(val) {
                if (!val) return

                this.$nextTick(() => {
                    this.buildDropdown()
                })
            },
        },
    }

</script>

<style lang="scss" scoped>

 @import '../../../resources/sass/variables/_all.scss';
    
    .switcher {
        display:block;
        height:40px;
        right:10px;
        margin-top: 5px;

        * {
            vertical-align: middle;
        }

        .switcher .scrolling .item {
            border-bottom: 1px solid rgba(34,36,38,.05);
        }

        .item {
            padding: 0 40px 0 0;
            min-width: 280px
        }

        .header {
            min-width: 280px;
            min-height:35px;
            padding: 0;

            .ui.avatar {
                margin: 0 1.1em 0 -3px !important;
                display: inline-block;
                width: 2em;
                height: 2em;
                border-radius: 500rem;
                font-size: $croud-font-size-medium;
            }

            .item.selected {
                background: $croud-colour-white;
            }
        }
    }

    .ui.dropdown .menu > .header {
        font-size: $croud-font-size-medium;
    }

    .item .label,
    .header .label {
        margin-left:35px;
        margin-right:130px;
        padding-top: 8px !important;
    }

    .item .label {
        margin-top:0;
        padding-top: 6px !important;
    }

    .header-action {
        position: absolute;
        z-index: $croud-layer-2;
        right: 10px;
        top: 12px;
    }

    .avatar {
        position:absolute;
    }

</style>
