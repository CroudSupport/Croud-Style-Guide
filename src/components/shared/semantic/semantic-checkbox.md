### semantic-checkbox
For our checkboxes, we use the semantic-checkbox component from the [croud-vue-semantic](https://github.com/CroudSupport/vue-semantic) library. Below are a few examples of semantic-checkbox in action. Click [here](http://croudsupport.github.io/vue-semantic/#semantic-checkbox) for full component docs.

### Basic usage

    <semantic-checkbox v-model="checkbox" label="🍕 Pizza" />

### Styling
Pass through the **type** prop to change the style of checkbo

    <div class="ui form">
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Checkbox" />
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Slider" type="slider"/>
        </div>
        <div class="ui field">
            <semantic-checkbox v-model="checkbox" label="Toggle" type="toggle" />
        </div>
    </div>

### Disabled

    <semantic-checkbox label="🍕 Pizza" type="toggle" :disabled="true"/>
