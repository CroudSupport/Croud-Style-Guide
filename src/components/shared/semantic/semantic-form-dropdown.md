### semantic-form-dropdown
For our form dropdowns, we use the semantic-form-dropdown component from the [croud-vue-semantic](https://github.com/croudtech/vue-semantic) library. Below are a few examples of semantic-form-dropdown in action. Click [here](http://croudtech.github.io/vue-semantic/#semantic-form-dropdown) for full component docs.

### Basic usage

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]"></semantic-form-dropdown>

### Styling
You can use the props to change the styling of the dropdown

    <div class="ui form">
        <div class="field">
            <label>Fluid</label>
            <semantic-form-dropdown fluid v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>

        <div class="three fields">
            <div class="field">
                <label>Search</label>
                <semantic-form-dropdown search v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Multiple</label>
                <semantic-form-dropdown multiple v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
            <div class="field">
                <label>Disabled</label>
                <semantic-form-dropdown :disabled="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
            </div>
        </div>

        <div class="inline field">
            <label>Inline</label>
            <semantic-form-dropdown :inline="true" v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" />
        </div>
    </div>

### Using slots
The **default** slot allows you to add custom options to the top of the dropdown

    <semantic-form-dropdown v-model="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]">
        <div class="item" v-data="3">
            <b>Baz</b>
        </div>
    </semantic-form-dropdown>

### Handle Events
This component emits the **dropdown-selected** event when an item is selected.

    <semantic-form-dropdown :value="dropdownVal" :options="[{id: 1, name: 'Foo'},{id: 2, name: 'Bar'}]" @dropdown-selected="$toasted.success('TaDa!!!')"></semantic-form-dropdown>
