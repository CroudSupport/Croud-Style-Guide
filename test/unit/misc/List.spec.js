import Vue from 'vue'
import List from '../../../src/components/shared/misc/List'
import '../../../semantic/dist/semantic'

Vue.use(require('croud-vue-semantic'))

const Constructor = Vue.extend(List)

const vm = new Constructor({
    propsData: {
        loading: false,
        headerItem: { name: 'foo' },
        listGetter: [
          { name: 'foo' },
          { name: 'bar' },
          { name: 'baz' },
        ],
    },
}).$mount()

describe('List', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })

    describe('loading', () => {
        beforeEach(() => {
            vm.loading = true
        })

        it('should match the snapshot', () => {
            expect(vm.$el).toMatchSnapshot()
        })
    })

    describe('hide header', () => {
        beforeEach(() => {
            vm.showHeader = false
        })

        it('should match the snapshot', () => {
            expect(vm.$el).toMatchSnapshot()
        })
    })

    describe('slots', () => {
        describe('header-action', () => {
            it('should match the snapshot', () => {
                const comp = new Vue({
                    components: { List },
                    render(h) {
                        return h('list', [
                            h('div', {
                                slot: 'header-action',
                            }, [
                                h('button', {
                                    class: 'ui button tiny blue',
                                    domProps: {
                                        innerHTML: 'test',
                                    },
                                }),
                            ]),
                        ])
                    },
                }).$mount()

                expect(comp.$el).toMatchSnapshot()
            })
        })
    })
})
