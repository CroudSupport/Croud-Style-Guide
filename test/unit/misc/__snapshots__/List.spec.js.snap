// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List hide header should match the snapshot 1`] = `
<div
  class="universal-list ui left top floating pointing dropdown tiny positive button"
  tabindex="0"
>
  <span
    class="text"
  >
    Switch
  </span>
  <!--  -->
  <div
    class="menu"
    tabindex="-1"
  >
    <!--  -->
    <!--  -->
    <!--  -->
    <div
      class="scrolling menu"
    >
      <div
        class="item"
      >
        foo
      </div>
      <div
        class="item"
      >
        bar
      </div>
      <div
        class="item"
      >
        baz
      </div>
      <!--  -->
    </div>
    <div
      class="ui inverted dimmer active"
    >
      <div
        class="ui text loader large"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;

exports[`List loading should match the snapshot 1`] = `
<div
  class="universal-list ui left top floating pointing dropdown tiny positive button"
  tabindex="0"
>
  <span
    class="text"
  >
    Switch
  </span>
  <!--  -->
  <div
    class="menu"
    tabindex="-1"
  >
    <!--  -->
    <div
      class="header"
    >
      <div
        class="item"
      >
        foo
      </div>
      <!--  -->
    </div>
    <div
      class="divider"
    />
    <div
      class="scrolling menu"
    >
      <div
        class="item"
      >
        foo
      </div>
      <div
        class="item"
      >
        bar
      </div>
      <div
        class="item"
      >
        baz
      </div>
      <!--  -->
    </div>
    <div
      class="ui inverted dimmer active"
    >
      <div
        class="ui text loader large"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;

exports[`List should match the snapshot 1`] = `
<div
  class="universal-list ui left top floating pointing dropdown tiny positive button"
  tabindex="0"
>
  <span
    class="text"
  >
    Switch
  </span>
  <!--  -->
  <div
    class="menu"
    tabindex="-1"
  >
    <!--  -->
    <div
      class="header"
    >
      <div
        class="item"
      >
        foo
      </div>
      <!--  -->
    </div>
    <div
      class="divider"
    />
    <div
      class="scrolling menu"
    >
      <div
        class="item"
      >
        foo
      </div>
      <div
        class="item"
      >
        bar
      </div>
      <div
        class="item"
      >
        baz
      </div>
      <!--  -->
    </div>
    <div
      class="ui inverted dimmer"
    >
      <div
        class="ui text loader large"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;

exports[`List slots header-action should match the snapshot 1`] = `
<div
  class="universal-list ui left top floating pointing dropdown tiny positive button"
  tabindex="0"
>
  <span
    class="text"
  >
    Switch
  </span>
  <!--  -->
  <div
    class="menu"
    tabindex="-1"
  >
    <div
      class="header-action"
    >
      <div>
        <button
          class="ui button tiny blue"
        >
          test
        </button>
      </div>
    </div>
    <div
      class="header"
    >
      <div
        class="item active selected"
      >
        
      </div>
      <div
        class="item"
      >
        Select an item
      </div>
    </div>
    <div
      class="divider"
    />
    <div
      class="scrolling menu"
    >
      <div
        class="item"
      >
        No items to display
      </div>
    </div>
    <div
      class="ui inverted dimmer"
    >
      <div
        class="ui text loader large"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;
